src/Razor/src/Microsoft.AspNetCore.Razor.Language/Intermediate/HtmlContentIntermediateNodeExtensions.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language.Intermediate
+{
+    public static class HtmlContentIntermediateNodeExtensions
+    {
+        private static readonly string HasEncodedContent = "HasEncodedContent";
+
+        public static bool IsEncoded(this HtmlContentIntermediateNode node)
+        {
+            return ReferenceEquals(node.Annotations[HasEncodedContent], HasEncodedContent);
+        }
+
+        public static void SetEncoded(this HtmlContentIntermediateNode node)
+        {
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            node.Annotations[HasEncodedContent] = HasEncodedContent;
+        }
+
+        public static string GetHtmlContent(this HtmlContentIntermediateNode node)
+        {
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            var builder = new StringBuilder();
+            var htmlTokens = node.Children.OfType<IntermediateToken>().Where(t => t.IsHtml);
+            foreach (var htmlToken in htmlTokens)
+            {
+                builder.Append(htmlToken.Content);
+            }
+            return builder.ToString();
```

Comment:
This is going to allocate a bunch... what would you think about just inlining the logic from `VisitHtml`